### ISSUE TITLE
Interrupting / cancelling running queries leads to memory leaks

### ISSUE THREAD

**Issue by FiV0** [xtdb/crux#2770]
Running the following test a couple of times eventually leads to memory leaks.
```clj
(deftest break-node

    (let [node-dir (io/file "dev/break-node")
          node-opts {:node-dir (.toPath node-dir)
                     :instant-src InstantSource/SYSTEM}]
      
      (with-open [node (tu/->local-node node-opts)]
        (doseq [tx (->> (range 100000)
                        shuffle
                        (map (comp #(vector :put :xt-docs %) #(hash-map :xt/id % :s (str %))))
                        (partition-all 1024))]
          (xt/submit-tx node tx)))
      
      (with-open [node (tu/->local-node node-opts)]

        (let [f (future (xt/q node '{:find [i] :where [(match :xt-docs {:xt/id i})]}))]
          (Thread/sleep 300)
          (future-cancel f)

          (t/is (true? (future-cancelled? f)))))))
```
I suspect that we are not properly closing resources in the TrieCursor when an async query gets interrupted.

**Comment by FiV0** [comment]
I confirmed that this issue is different to #2544. I also get memory leak errors when commenting out the first `with-open` on subsequent runs.

**Comment by FiV0** [comment]
I suspect there is still some issue with async queries that is harder to reproduce but I think the initial main reason this test broke was that we were not closing the `ra-query-source` properly which was fixed in 80236da7.
